File: Sintaxis I/_10 Scope.js
// ═════════════════════════════════════
//                Scope
// ═════════════════════════════════════

// Scope define dónde una variable está disponible

// ┌───────────────────────────────────┐
// │      1. Scope Global              │
// └───────────────────────────────────┘
const globalVar = 'Soy global' // Accesible en todo el archivo
function verGlobal () {
  console.log(globalVar) // "Soy global" (visible aquí)
}
verGlobal()

// ┌───────────────────────────────────┐
// │      2. Scope Local (Function)    │
// └───────────────────────────────────┘
function scopeLocal () {
  const localVar = 'Soy local' // Solo existe dentro de esta función
  console.log(localVar) // "Soy local"
}
scopeLocal()
// console.log(localVar);        // Error: localVar no está definida fuera

// ┌───────────────────────────────────┐
// │      3. Scope de Bloque           │
// └───────────────────────────────────┘
if (true) {
  const bloqueVar = 'Soy de bloque' // Solo vive en este bloque con let/const
  var noBloque = 'No respeto bloques' // var ignora bloques, sube al scope superior
  console.log(bloqueVar) // "Soy de bloque"
}
// console.log(bloqueVar);              // Error: no accesible fuera del bloque
console.log(noBloque) // "No respeto bloques" (var es global o funcional)

// ┌───────────────────────────────────┐
// │      4. Anidamiento               │
// └───────────────────────────────────┘
function exterior () {
  const externa = 'Estoy afuera'
  function interior () {
    const interna = 'Estoy adentro'
    console.log(externa) // "Estoy afuera" (accede al scope superior)
    console.log(interna) // "Estoy adentro"
  }
  interior()
  // console.log(interna);  // Error: interna no está disponible aquí
}
exterior()

// ┌───────────────────────────────────┐
// │      5. let vs var vs const       │
// └───────────────────────────────────┘
const vieja = 'Soy var' // Scope funcional, redeclarable
let moderna = 'Soy let' // Scope de bloque, reasignable
const fija = 'Soy const' // Scope de bloque, no reasignable
moderna = 'Cambié' // Ok
// fija = "No puedo";     // Error: const no permite reasignación

console.log(vieja, moderna, fija)

File: Sintaxis I/_10 Scope.py
# ┌────────────────────────
# │  => SCOPE - I
# └────────────────────────

#  Global
x = "awesome"


def myfunc():
    #  Local
    x = "fantastic"
    print("Python is " + x)


myfunc()  # -- "Python is  fantastic"

print("Python is " + x)  # -- "Python is  awesome" [# No Sobreescribir Global]

# ┌────────────────────────
# │  => SCOPE - II
# └────────────────────────


def myfunc():
    global x  # -- Creacion de Una Variable GLOBAL
    x = "fantastic"


myfunc()

print("Python is " + x)

# ┌────────────────────────
# │  => SCOPE - III
# └────────────────────────

#  Global
x = "awesome"


def myfunc():
    global x  # --Sobreescribir un Variable Global [global]
    x = "fantastic"


myfunc()

print("Python is " + x)

# ┌────────────────────────
# │  => SCOPE - III
# └────────────────────────

x = 300


def myfunc():
    x = 200
    print(x)


myfunc()  # 200

print(x)  # 300

x = 500


def myfunc():
    global x
    x = 300


myfunc()

print(x)  # 300
